crypto/provider_core.c

```diff
@@ -0,0 +1,410 @@
+    OPENSSL_free(store);
+}
+
+static void *provider_store_new(void)
+{
+    struct provider_store_st *store = OPENSSL_zalloc(sizeof(*store));
+
+    if (store == NULL
+        || (store->providers = sk_OSSL_PROVIDER_new(ossl_provider_cmp)) == NULL
+        || (store->lock = CRYPTO_THREAD_lock_new()) == NULL) {
+        provider_store_free(store);
+        store = NULL;
+    }
+    return store;
+}
+
+static const OPENSSL_CTX_METHOD provider_store_method = {
+    provider_store_new,
+    provider_store_free,
+};
+
+static CRYPTO_ONCE provider_store_init_flag = CRYPTO_ONCE_STATIC_INIT;
+DEFINE_RUN_ONCE_STATIC(do_provider_store_init)
+{
+    return OPENSSL_init_crypto(0, NULL)
+        && (provider_store_index =
+            openssl_ctx_new_index(&provider_store_method)) != -1;
+}
+
+
+static struct provider_store_st *get_provider_store(OPENSSL_CTX *libctx)
+{
+    if (!RUN_ONCE(&provider_store_init_flag, do_provider_store_init))
+        return NULL;
+    return openssl_ctx_get_data(libctx, provider_store_index);
+}
+
+/*-
+ * Provider Object methods
+ * =======================
+ */
+
+int ossl_provider_upref(OSSL_PROVIDER *prov)
+{
+    int ref = 0;
+
+#ifndef HAVE_ATOMICS
+    CRYPTO_UP_REF(&prov->refcnt, &ref, prov->refcnt_lock);
+#else
+    CRYPTO_UP_REF(&prov->refcnt, &ref, NULL);
+#endif
+    return ref;
+}
+
+/* Finder, constructor and destructor */
+OSSL_PROVIDER *ossl_provider_find(OPENSSL_CTX *libctx, const char *name)
+{
+    struct provider_store_st *store = NULL;
+    OSSL_PROVIDER *prov = NULL;
+    OSSL_PROVIDER tmpl = { 0, };
+
+    tmpl.name = (char *)name;
+    if ((store = get_provider_store(libctx)) != NULL) {
+        int i;
+
+        CRYPTO_THREAD_write_lock(store->lock);
+        if ((i = sk_OSSL_PROVIDER_find(store->providers, &tmpl)) == -1
+            || (prov = sk_OSSL_PROVIDER_value(store->providers, i)) == NULL
+            || !ossl_provider_upref(prov))
+            prov = NULL;
+        CRYPTO_THREAD_unlock(store->lock);
+    }
+
+    return prov;
+
+}
+
+OSSL_PROVIDER *ossl_provider_new(OPENSSL_CTX *libctx, const char *name,
+                                 ossl_provider_init_fn *init_function)
+{
+    struct provider_store_st *store = get_provider_store(libctx);
+    OSSL_PROVIDER *prov = NULL;
+
+    if ((prov = ossl_provider_find(libctx, name)) != NULL) { /* refcount +1 */
+        ossl_provider_free(prov); /* refcount -1 */
+        CRYPTOerr(CRYPTO_F_OSSL_PROVIDER_NEW,
+                  CRYPTO_R_PROVIDER_ALREADY_EXISTS);
+        ERR_add_error_data(2, "name=", name);
+        return NULL;
+    }
+
+    if ((prov = OPENSSL_zalloc(sizeof(*prov))) == NULL
+        || !ossl_provider_upref(prov) /* +1 One reference to be returned */
+        || (prov->name = OPENSSL_strdup(name)) == NULL) {
+        ossl_provider_free(prov);
+        CRYPTOerr(CRYPTO_F_OSSL_PROVIDER_NEW, ERR_R_MALLOC_FAILURE);
+        return NULL;
+    }
+
+    prov->init_function = init_function;
+
+    CRYPTO_THREAD_write_lock(store->lock);
```

Comment:
Hmm - not good if store = NULL (which is checked 2 lines down)

Reply:
Do we really want to cater for the store == NULL. Could that just be a error?

Reply:
That is an error on my part...  I wasn't attentive when moving code around at some point...